### ISSUE TITLE
add support for async/await and eventloop instead of threads for async stdin

### ISSUE THREAD

**Issue by prabirshrestha** [TimonPost/crossterm#91]
Not sure about this but curious if async event loop would be better then threads similar to alacritty https://github.com/jwilm/alacritty/blob/3c672cca4b469d1e9e59db819fe459358522f6ec/src/tty/mod.rs

**Comment by TimonPost** [comment]
Interesting I will take a look into this. I agree that the threads are probably not the best here.

**Comment by imdaveho** [comment]
Is there a reason that async / await would be better? I am intending to use crossterm as a library to wrap Python. Threads seems to be a more cross platform compatible mechanism to achieve concurrency. I could be wrong though, but I'm also not a huge fan of eventloops where languages already provide you safety guarantees with threads...is there really a benefit that I'm not aware of?

**Comment by TimonPost** [comment]
Agreed, bringing in event loops, async await futures might increase the complexity of crossterm. The one thread we have will exist for a little time and will be closed after that. I think it is not much needed to bring more complexity in.

**Comment by prabirshrestha** [comment]
The main reason for me was if we could avoid thread sleep here and avoid wasting CPU cycles sepcially on laptops.

https://github.com/TimonPost/crossterm/blob/c37a480f78c8c1ca020271ae6902d453122955e9/crossterm_input/examples/async_input.rs#L36

My thought was that using async/await if there is some sort of `mpsc::Receiver` or `Streams` then internally it doesn't matter whether it using thread of eventloop sine the consumer of the api wouldn't need to worry about changing the api.

```
let input_stream = input().read_async_with_threads(); // or input().read_async_witheventloop();
while let Some(inputEvent) = await!(input_stream().next()) {
    // do something with inputEvent
}
```

As for the actual perf I'm not sure which one is better but in terms of battery life I think eventloops are better.

You do bring up a good point where threads are more portable so it does seems good to have that as an option.

We should wait for generators/streams to finalize first.

**Comment by TimonPost** [comment]
Agree that async and await might be better however it is does not have good support yet in rust. If we want to use async await need to use Tokio/futures or something which I like to stay far a way from with crossterm. Maybe 'awaiting' for rust to introduce async and await is the best idea here.

**Comment by prabirshrestha** [comment]
I agree we should wait for rust to stabilize async await first.

**Comment by TimonPost** [comment]
This could be closed, we stick with threads, #79 is fixed which kind of relates to this.
